README.md
[<p align="center"><img src="doc/logo.png" alt="node-apn" width="450" height="auto"></p>][node-apn]

node-apn
========

> A Node.js module for interfacing with the Apple Push Notification service.

[![Maintenance help wanted](https://img.shields.io/badge/maintenance-help%20wanted-red.svg)](https://github.com/node-apn/node-apn/issues/436)
[![Build status][ci-image] ][ci-url]
[![Code coverage][coverage-image]][coverage-url]
[![Codacy][codacy-image]][codacy-url]
[![dependencies][dependencies-image]][dependencies-url]
[![devdependencies][devdependencies-image]][devdependencies-url]

[logo]:doc/logo.png
[npm-image]:https://nodei.co/npm/apn.png?downloads=true
[npm-url]:https://npmjs.com/package/apn
[ci-image]:https://api.travis-ci.org/node-apn/node-apn.svg
[ci-url]:https://travis-ci.org/node-apn/node-apn
[coverage-image]:https://coveralls.io/repos/argon/node-apn/badge.svg?branch=develop
[coverage-url]:https://coveralls.io/r/argon/node-apn
[codacy-image]:https://www.codacy.com/project/badge/e7735fbe0db244f3b310657d0dabaa11
[codacy-url]:https://www.codacy.com/public/argon/node-apn

[dependencies-image]:https://david-dm.org/node-apn/node-apn/status.svg
[dependencies-url]:https://david-dm.org/node-apn/node-apn
[devdependencies-image]:https://david-dm.org/node-apn/node-apn/dev-status.svg
[devdependencies-url]:https://david-dm.org/node-apn/node-apn?type=dev

## Features

- Based on HTTP/2 based provider API
- Maintains a connection to the server to maximise notification batching and throughput.
- Automatically resends unsent notifications if an error occurs

## Installation

[npm][] is the preferred installation method:

```bash
$ npm install apn --save
```

## Quick Start

This readme is a brief introduction, please refer to the full [documentation](doc/apn.markdown) in `doc/` for more details.

If you have previously used v1.x and wish to learn more about what's changed in v2.0, please see [What's New](doc/whats-new.markdown)

### Load in the module

```javascript
var apn = require('apn');
```

### Connecting
Create a new connection to the Apple Push Notification provider API, passing a dictionary of options to the constructor. You must supply your token credentials in the options.

```javascript
var options = {
  token: {
    key: "path/to/APNsAuthKey_XXXXXXXXXX.p8",
    keyId: "key-id",
    teamId: "developer-team-id"
  },
  production: false
};

var apnProvider = new apn.Provider(options);
```

By default, the provider will connect to the sandbox unless the environment variable `NODE_ENV=production` is set.

For more information about configuration options consult the [provider documentation](doc/provider.markdown).

Help with preparing the key and certificate files for connection can be found in the [wiki][certificateWiki]

#### Connecting through an HTTP proxy

If you need to connect through an HTTP proxy, you simply need to provide the `proxy: {host, port}` option when creating the provider. For example:

```javascript
var options = {
  token: {
    key: "path/to/APNsAuthKey_XXXXXXXXXX.p8",
    keyId: "key-id",
    teamId: "developer-team-id"
  },
  proxy: {
    host: "192.168.10.92",
    port: 8080
  },
  production: false
};

var apnProvider = new apn.Provider(options);
```

The provider will first send an HTTP CONNECT request to the specified proxy in order to establish an HTTP tunnel. Once established, it will create a new secure connection to the Apple Push Notification provider API through the tunnel.

### Sending a notification
To send a notification you will first need a device token from your app as a string

```javascript
let deviceToken = "a9d0ed10e9cfd022a61cb08753f49c5a0b0dfb383697bf9f9d750a1003da19c7"
```

Create a notification object, configuring it with the relevant parameters (See the [notification documentation](doc/notification.markdown) for more details.)

```javascript
var note = new apn.Notification();

note.expiry = Math.floor(Date.now() / 1000) + 3600; // Expires 1 hour from now.
note.badge = 3;
note.sound = "ping.aiff";
note.alert = "\uD83D\uDCE7 \u2709 You have a new message";
note.payload = {'messageFrom': 'John Appleseed'};
note.topic = "<your-app-bundle-id>";
```

Send the notification to the API with `send`, which returns a promise.

```javascript
apnProvider.send(note, deviceToken).then( (result) => {
  // see documentation for an explanation of result
});
```

This will result in the the following notification payload being sent to the device

```json
{"messageFrom":"John Appleseed","aps":{"badge":3,"sound":"ping.aiff","alert":"\uD83D\uDCE7 \u2709 You have a new message"}}
```

You should only create one `Provider` per-process for each certificate/key pair you have. You do not need to create a new `Provider` for each notification. If you are only sending notifications to one app then there is no need for more than one `Provider`. 

If you are constantly creating `Provider` instances in your app, make sure to call `Provider.shutdown()` when you are done with each provider to release its resources and memory.

## Troubleshooting

You are encouraged to read the extremely informative [Troubleshooting Push Notifications][tn2265] Tech Note in the first instance, in case your query is answered there.

## Support

If you have any questions or difficulties working with the module, the [node-apn Google group][googlegroup] should be your first port of call.

Please include as much detail as possible - especially debug logs. If the problem is reproducible, sample code is also extremely helpful. GitHub Issues should only be created for verified problems and enhancements, this will allow them to be tracked more easily.

## Resources

* [Local and Push Notification Programming Guide: APNs Overview][pl]
* [Apple Technical Note: Troubleshooting Push Notifications][tn2265]
* [List of Projects, Applications and Companies Using Node-apn][pacapn]

## Credits

Created by [Andrew Naylor][argon]

Thanks to: [Ian Babrou][bobrik], [dgthistle][dgthistle], [Keith Larsen][keithnlarsen], [Mike P][mypark], [Greg Bergé][neoziro], [Asad ur Rehman][AsadR], [Nebojsa Sabovic][nsabovic], [Alberto Gimeno][gimenete], [Randall Tombaugh][rwtombaugh], [Michael Stewart][thegreatmichael], [Olivier Louvignes][mgcrea], [porsager][porsager], [Craig Hockenberry][chockenberry]

## License

Released under the MIT License

> Copyright (c) 2013 Andrew Naylor
>
> Permission is hereby granted, free of charge, to any person obtaining a copy
> of this software and associated documentation files (the "Software"), to deal
> in the Software without restriction, including without limitation the rights
> to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
> copies of the Software, and to permit persons to whom the Software is
> furnished to do so, subject to the following conditions:
>
> The above copyright notice and this permission notice shall be included in
> all copies or substantial portions of the Software.
>
> THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

[npm]: https://npmjs.org
[node-apn]: https://github.com/node-apn/node-apn
[certificateWiki]:https://github.com/node-apn/node-apn/wiki/Preparing-Certificates "Preparing Certificates"
[pl]: https://developer.apple.com/library/content/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/APNSOverview.html "Local and Push Notification Programming Guide: APNs Overview"
[tn2265]: http://developer.apple.com/library/ios/#technotes/tn2265/_index.html "Troubleshooting Push Notifications"
[googlegroup]:https://groups.google.com/group/node-apn "node-apn Google Group"
[pacapn]:https://github.com/node-apn/node-apn/wiki/Projects,-Applications,-and-Companies-Using-Node-apn "List of Projects, Applications and Companies Using node-apn"
[argon]: https://github.com/argon
[bobrik]: https://github.com/bobrik
[dgthistle]: https://github.com/dgthistle
[keithnlarsen]: https://github.com/keithnlarsen
[mypark]: https://github.com/mypark
[neoziro]: https://github.com/neoziro
[AsadR]: https://github.com/AsadR
[nsabovic]: https://github.com/nsabovic
[gimenete]: https://github.com/gimenete
[rwtombaugh]: https://github.com/rwtombaugh
[thegreatmichael]: https://github.com/thegreatmichael
[mgcrea]: https://github.com/mgcrea
[porsager]: https://github.com/porsager
[chockenberry]: https://github.com/chockenberry



ChangeLog.md
## Changelog
[Full Changelog](https://github.com/node-apn/node-apn/compare/v3.0.0...master)

3.0.0: [Full ChangeLog](https://github.com/node-apn/node-apn/compare/v2.2.0...3.0.0)
* Support zero-valued (0) `expiry`[#649](https://github.com/node-apn/node-apn/pull/649).  Thanks to [Nathan Hamblen](https://github.com/n8han)
* add support for apns-push-type [#656](https://github.com/node-apn/node-apn/pull/656).  Thanks to [Stefan Trauth](https://github.com/funkenstrahlen)

2.2.0:
* Proxy support (#163)
* Updated dependencies

2.1.5:
* Improve typescript type definitions (#536)
* Several documentation improvements (#537, #547, #550, #555, #560, #560)
* Update dependencies (#566)
* Update .gitignore and add .nvmrc (#573)

2.1.4:
* Don't mutate passed in payload object
* Change APNs development endpoint to api.development.push.apple.com
* Update node-http2 fork

2.1.3:
* Add thread-id to notification payload (#484)
* Workaround an issue where APNS would return an HTTP 500 and make the connection unusable (#449, #480)
* Regularily renew JSON web token (#449, #480)

2.1.2:
* Add TypeScript type definitions (#441, #465)
* Update node-http2 fork to 3.3.6 and add stream.upstream not defined check (#444)

2.1.1:
* Fix broken logger (#438)
* Use archive URL for forked node-http2 module (#433)

2.1.0:

* JSON Web Token support (#401)
* Add `threadId` property for notifications
* Fix a leak caused when an error is emitted from the endpoint manager
    - Queued notifications would resolve as errored but would remain in the queue, leading them to be repeatedly rejected without termination and causing an error log from node

2.0.0:

* HTTP/2 Provider API (#277)
* `apn.Connection` has been renamed to `apn.Provider`
* `apn.Feedback` has been removed
* `apn.Device` has been removed - all tokens are now hex-encoded strings
* `apn.token` is provided to validate tokens and convert from `Buffer` if
  necessary
* Notifications are now required to have an associated `topic`
* `pushNotification(notification, tokens)` is now simply, `send(notification, recipients)`
* `send` returns a promise which will be fulfilled when all notifications have
  been sent (#362)
* Reliable Delivery (#168)
* Ability to mock `apn.Provider` behaviour
* Notification can be created in a "finished" state (#240)

1.7.8:

* Add support for Mutable Content (#413)

1.7.7:

* Fix Documentation URL (#393)
* Prevent `badge: 0` sending `badge: undefined` (#394)
* Avoid error, when deferredConnection is null (#397)

1.7.6:

* Emit an error when retry limit is exceeded (#333)
* Documentation fixes (#229, #343, #379)
* Reinstate broken aps behaviour (#377)

1.7.5:

* Notification property improvements
* Round-robin all open sockets when sending notifications
* `minConnections` option
* Prevent crashing when cancelling feedback multiple times (#307)
* Fix connection.shutdown behaviour (#299)
* Small doc improvements

1.7.4:

* Notification trim fixes. (#220)
* Notification refactoring
* `debug` package is now a dependency (#267)

1.7.3:

* Added support for increased payload length for VoIP applications (Closes #207)
* Fixed a bug with trimming UTF-16 encoded payloads
* Dropped support for node v0.6 as it doesn't support UTF-16 surrogate pairs. Plus it's old. It'll still work if needed though, if you use UTF-8.

1.7.2:

* Fixed: #238, only emit `error` when the problem is unrecoverable. Any use of `node-apn` should have an `error` listener attached to prevent uncaught exceptions.
* Various coding style improvements.
* CI.
* Removed legacy protocol support. I don't see any reason to keep it around, let me know if this causes you any problems.

1.7.1:

* Fixed: #224, always passing a CA value even if no certificates were specified. In this case the TLS library will not use the built in root certificates and will always fail to trust the server certificate.
* Changed: Socket timeout default has changed for disabled to 1 hour.
* Documentation fixes

1.7.0:

* Added: Credential validator to catch common configuration errors.
* Fixed: Documentation errors.

1.6.2:

* Updated maximum payload size to 2048 bytes. (See #181).
* Fixed: Feedback wouldn't include error message when emitting `feedbackError`.

1.6.1:

* Increased default cache length to 1000; Previous value of 100 was overly cautious.
* Added a delay of 100ms to connection to ensure notifications aren't lost if wrong certificates are used.
* Fixes #195: Better handling of socket creation
* Fixes #200: Improved buffer handling code when multiple connections are enabled.
* Minor optimisation to Notification processing, removing the need for 2 `JSON.stringify` calls
* Fixes #196: Check cache has contents before trying to access. Also ensure minimum size is "1" to allow "transmitted" events to be emitted.
* Fixes #199: Emit a "drain" event when no further notifications require sending. Most useful in a batch environment.

1.6.0:

* Fixes #156: Retire a socket from the pool immediately instead of waiting for a "close" event.
* Fixes #165: Introduced a connection retry limit to emit errors when fatal connection errors occur.
* Fixes #152: Failure to attach an error handler would prevent exponential backoff from working.
* Improved error handling for EPIPE write events to maximise chances of reading error info from Apple.
* Fixed #181: Initial support for larger payloads for iOS 8 testing.
* Fixed #187: Correctly trim notifications with escaped characters.
* Merged #190: Always emit a feedback event when batch mode is enabled, even if there are no tokens.
* Merged #189: Added support for iOS 8 Notification actions.
* Display a warning when null parameters are passed during connection creation.
* Fixed #171: Removed references to the unknown notification identifier.

1.5.2:

* Fixed #169, #170: Undesirable behaviour when PFX files are specified
* Fixes several problems identified after adding further test coverage.

1.5.0/1.5.1:


lib/apn.js
var tls = require('tls');
var fs = require('fs');
var Buffer = require('buffer').Buffer;

var Errors = {
	'noErrorsEncountered': 0,
	'processingError': 1,
	'missingDeviceToken': 2,
	'missingTopic': 3,
	'missingPayload': 4,
	'invalidTokenSize': 5,
	'invalidTopicSize': 6,
	'invalidPayloadSize': 7,
	'invalidToken': 8,
	'none': 255
};

var Connection = function (optionArgs) {
	var currentId = 0;
	var cachedNotes = [];

	var self = this;
	var hasKey = false;
	var hasCert = false;
	var socketOptions = {};
	var openingSocket = false;
	var writeBuffer = [];

	var options = {
		cert: 'cert.pem' /* Certificate file */,
		certData: '',
		key: 'key.pem'	/* Key file */,
		keyData: '',
		passphrase: null,
		gateway: 'gateway.push.apple.com' /* gateway address */,
		port: 2195 /* gateway port */,
		enhanced: true /* enable enhanced format */,
		errorCallback: undefined /* Callback when error occurs */,
		cacheLength: 5 /* Number of notifications to cache for error purposes */
	};
	
	if (optionArgs) {
		var keys = Object.keys(options);
		for (var i = 0; i < keys.length; i++) {
			var k = keys[i];
			if (optionArgs[k] !== undefined) options[k] = optionArgs[k];
		}
	}

	var readyToConnect = function () {
		return (hasKey && hasCert);
	};

	var onDrain = function () {
		while (writeBuffer.length && self.socket.socket.bufferSize == 0) {
			writeNotificationToSocket(writeBuffer.shift());
		}
	};

	var startSocket = function () {
		if (!self.openingSocket) {
			process.nextTick(function () {
				self.socket = tls.connect(options['port'], options['gateway'], socketOptions,
					function () {
						if (!self.socket.authorized) {
							throw self.socket.authorizationError;
						}

						onDrain();
						self.openingSocket = false;
					});

				self.socket.on('data', function (data) {
					handleTransmissionError(data);
				});

				self.socket.on('error', function () {
					self.socket.removeAllListeners();
					self.socket = undefined;
				});

				self.socket.once('close', function () {
					if (writeBuffer.length && readyToConnect()) {
						startSocket();
					}
					if (!self.socket) {
						return;
					}

					self.socket.removeAllListeners();
					self.socket = undefined;
				});

				self.socket.on("drain", onDrain);
			});
			self.openingSocket = true;
		}
	};

	var connect = invoke_after(function () {
		startSocket();
	});

	if (options['certData']) {
		socketOptions['cert'] = options['certData'];
		hasCert = true;
	} else {
		fs.readFile(options['cert'], connect(function (err, data) {
			if (err) {
				throw err;
			}
			socketOptions['cert'] = data.toString();
			hasCert = true;
		}));
	}

	if (options['keyData']) {
		socketOptions['key'] = options['keyData'];
		hasKey = true;
	} else {
		fs.readFile(options['key'], connect(function (err, data) {
			if (err) {
				throw err;
			}
			socketOptions['key'] = data.toString();
			hasKey = true;
		}));
	}

	if (options['passphrase']) {
		socketOptions['passphrase'] = options['passphrase'];
	}

	var writeNotificationToSocket = function (data) {
		if (self.socket === undefined || self.socket.readyState != 'open') {
			if ((self.socket === undefined || self.socket.readyState == 'closed') && readyToConnect()) {
				startSocket();
			}
			bufferDataForWrite(data);
		} else {
			if (self.socket.socket.bufferSize > 0) {
				bufferDataForWrite(data);
			} else {
				self.socket.write(data);
			}
		}
	};

	var bufferDataForWrite = function (data) {
		writeBuffer.push(data);
	};

	this.sendNotification = function (note) {
		var encoding = 'utf8';
		if (note.encoding) {
			encoding = note.encoding;
		}
		var token = note.device.token;
		var message = JSON.stringify(note);
		var messageLength = Buffer.byteLength(message, encoding);
		var pos = 0;

		if (token === undefined) {
			return Errors['missingDeviceToken'];
		}
		if (messageLength > 256) {
			return Errors['invalidPayloadSize'];
		}

		note._uid = currentId++;
		var data;
		if (options.enhanced) {
			data = new Buffer(1 + 4 + 4 + 2 + token.length + 2 + messageLength);
			// Command
			data[pos] = 1;
			pos++;

			// Identifier
			pos += int2buf(note._uid, data, pos, 4);

			// Expiry
			pos += int2buf(note.expiry, data, pos, 4);

			cachedNotes.push(note);
			tidyCachedNotes();
		}
		else {
			data = new Buffer(1 + 2 + token.length + 2 + messageLength);
			data[pos] = 0;
			pos++;
		}

		pos += int2buf(token.length, data, pos, 2);
		pos += token.copy(data, pos, 0);
		pos += int2buf(messageLength, data, pos, 2);
		pos += data.write(message, pos, encoding);

		writeNotificationToSocket(data);
	};

	var tidyCachedNotes = function () {
		// Maybe a timestamp should be stored for each note and kept for a duration?
		if (cachedNotes.length > options.cacheLength) {
			cachedNotes.shift();
		}
	};

	var handleTransmissionError = function (data) {
		// Need to check message that errors
		// return failed notification to owner
		// purge writeBuffer and start again with cachedNotes
		//  perhaps we should keep an identifier with each writeBuffer
		//  entry to know where to stop when cycling through cachedNotes
		if (data[0] == 8) {

			if (!options.enhanced) {
				cachedNotes = [];
				return;
			}

			// This is an error condition
			var errorCode = data[1];
			var identifier = bytes2int(data.slice(2, 6), 4);
			var note = undefined;

			while (cachedNotes.length) {
				note = cachedNotes.shift();
				if (note['_uid'] == identifier) {
					break;
				}
			}

			// Notify callback of failed notification
			if (typeof options.errorCallback == 'function') {
				options.errorCallback(errorCode, note);
			}
			writeBuffer = [];
			var count = cachedNotes.length;
			for (var i = 0; i < count; i++) {
				note = cachedNotes.shift();
				self.sendNotification(note);
			}
		}
	}
};

var Notification = function () {
	this.payload = {};
	this.expiry = 0;
	this.identifier = 0;
	this.device;

	this.alert = undefined;
	this.badge = undefined;
	this.sound = undefined;
};

Notification.prototype.toJSON = function () {
	if (this.payload === undefined) {
		this.payload = {};
	}
	if (this.payload.aps === undefined) {
		this.payload.aps = {};
	}
	if (typeof this.badge == 'number') {
		this.payload.aps.badge = this.badge;
	}
	if (typeof this.sound == 'string') {
		this.payload.aps.sound = this.sound;
	}
	if (typeof this.alert == 'string' || typeof this.alert == 'object') {
		this.payload.aps.alert = this.alert;
	}
	return this.payload;
};
	
var Device = function (/* deviceToken, ascii=true */) {
	var self = this;
	self.token = undefined;

	if (arguments.length > 0) {
		self.setToken.apply(self, arguments);
	}
};
	
Device.prototype.parseToken = function (token) {
	token = token.replace(/\s/g, "");
	var length = Math.ceil(token.length / 2);
	var hexToken = new Buffer(length);
	for (var i = 0; i < token.length; i += 2) {
		var word = token[i];
		if ((i + 1) >= token.length || typeof(token[i + 1]) === undefined) {
			word += '0';
		}
		else {
			word += token[i + 1];
		}
		hexToken[i / 2] = parseInt(word, 16);
	}
	return hexToken;
};
	
Device.prototype.setToken = function (newToken, ascii) {
	if (ascii === undefined || ascii == true) {
		newToken = this.parseToken(newToken);
	}
	this.token = newToken;
	return this;
};
	
Device.prototype.hexToken = function () {
	var out = [];
	var len = this.token.length;
	var n;
	for (var i = 0; i < len; i++) {
		n = this.token[i];
		if (n < 16) out[i] = "0" + n.toString(16);
		else out[i] = n.toString(16);
	}
	return out.join("");
};

var Feedback = function (optionArgs) {
	var self = this;
	var hasKey = false;
	var hasCert = false;
	var socketOptions = {};

	var responsePacketLength = 38;
	var readBuffer = new Buffer(responsePacketLength);
	var readLength = 0;

	var options = {
		cert: 'cert.pem',					/* Certificate file */
		certData: '',						/* Certificate data */
		key: 'key.pem',						/* Key file */
		keyData: '',						/* Key data */
		passphrase: null,                   /* Passphrase for key */
		address: 'feedback.push.apple.com',	/* feedback address */
		port: 2196,							/* feedback port */
		feedback: false,					/* enable feedback service, set to callback */
		interval: 3600,						/* interval in seconds to connect to feedback service */
	};

	if (optionArgs) {
		var keys = Object.keys(options);
		for (var i = 0; i < keys.length; i++) {
			var k = keys[i];
			if (optionArgs[k] !== undefined) options[k] = optionArgs[k];
		}
	}

	if (typeof options['feedback'] != 'function') {
		return Error(-1, 'A callback function must be specified');
	}

	this.readyToConnect = function () {
		return (hasKey && hasCert);
	};

	self.startSocket = function () {
		self.socket = tls.connect(options['port'], options['address'], socketOptions);
		self.socket.pair.on('secure', function () {
			if (!self.socket.authorized) {
				throw self.socket.authorizationError;
			}
		});
		self.socket.on('data', function (data) {
			processData(data);
		});
		self.socket.once('error', function () {
			self.socket.removeAllListeners();
			self.socket = undefined;
		});
		self.socket.once('end', function () {
		});
		self.socket.once('close', function () {
			self.socket.removeAllListeners();
			self.socket = undefined;
		});
	};

	var connect = invoke_after(function () {
		self.startSocket();
	});

	if (options['certData']) {
		socketOptions['cert'] = options['certData'];
		hasCert = true;
	} else {
		fs.readFile(options['cert'], connect(function (err, data) {
			if (err) {
				throw err;
			}
			socketOptions['cert'] = data.toString();
			hasCert = true;
		}));
	}

	if (options['keyData']) {
		socketOptions['key'] = options['keyData'];
		hasKey = true;
	} else {
		fs.readFile(options['key'], connect(function (err, data) {
			if (err) {
				throw err;
			}
			socketOptions['key'] = data.toString();
			hasKey = true;
		}));
	}

	if (options['passphrase']) {
		socketOptions['passphrase'] = options['passphrase'];
	}

	if (options['interval'] > 0) {
		this.interval = setInterval(function () {
			self.request();
		}, options['interval'] * 1000);
	}

	var processData = function (data) {
		var pos = 0;
		// If there is some buffered data, read the remainder and process this first.
		if (readLength > 0) {
			if (data.length < (responsePacketLength - readLength)) {
				data.copy(readBuffer, readLength, 0);
				readLength += data.length;
				return;
			}
			data.copy(readBuffer, readLength, 0, responsePacketLength - readLength);
			decodeResponse(readBuffer, 0);
			pos = responsePacketLength - readLength;
			readLength = 0;
		}
		while (pos < data.length - 1) {
			if ((data.length - pos) < responsePacketLength) {
				//Buffer remaining data until next time
				data.copy(readBuffer, 0, pos);
				readLength = data.length - pos;
				break;
			}
			decodeResponse(data, pos);
			pos += responsePacketLength;
		}
	};
	
	var decodeResponse = function (data, start) {
		var time = bytes2int(data, 4, start);
		start += 4;
		var len = bytes2int(data, 2, start);
		start += 2;
		var tok = new Buffer(len);
		data.copy(tok, 0, start, start + len);
	
		if (typeof options['feedback'] == 'function') {
			options['feedback'](time, new exports.Device(tok, false));
		}
	}
};

Feedback.prototype.request = function () {
	if ((this.socket === undefined || this.socket.readyState == 'closed') && this.readyToConnect()) {
		this.startSocket();
	}
};

Feedback.prototype.cancel = function () {
	if (this.interval !== undefined) {
		clearInterval(this.interval);
	}
};

function int2buf (number, buffer, start, length) {
	length -= 1;
	for (var i = 0; i <= length; i++) {
		buffer[start + length - i] = number & 0xff;
		number = number >> 8;
	}
	return length + 1;
}

function bytes2int (bytes, length, start) {
	if (start === undefined) start = 0;
	var num = 0;
	length -= 1;
	for (var i = 0; i <= length; i++) {
		num += (bytes[start + i] << ((length - i) * 8));
	}
	return num;
}

function invoke_after (callback) {
	var n = 0;
	return function (delegate) {
		n++;
		return function () {
			delegate.apply(delegate, arguments);
			if (--n == 0) callback();
		};
	};
}

exports.Connection = Connection;
exports.connection = Connection;
exports.Notification = Notification;
exports.notification = Notification;
exports.Device = Device;
exports.device = Device;
exports.Feedback = Feedback;
exports.feedback = Feedback;
exports.error = Errors;

